src/utils/logging.hpp

```diff
@@ -22,31 +22,33 @@
 #include <filesystem>
 #include <optional>
 
+#include <fmt/format.h>
 #include <spdlog/fmt/ostr.h>
 #include <spdlog/sinks/stdout_color_sinks.h>
 #include <spdlog/spdlog.h>
 
+#include <boost/preprocessor/comparison/equal.hpp>
+#include <boost/preprocessor/control/if.hpp>
+#include <boost/preprocessor/variadic/size.hpp>
+
 namespace logging {
 
 // TODO (antonio2368): Replace with std::source_location when it's supported by
 // compilers
-template <typename... Args>
-void AssertFailed(const char *file_name, int line_num, const char *expr, const Args &...msg_args) {
-  std::optional<std::string> message;
-  if constexpr (sizeof...(msg_args) > 0) {
-    message.emplace(fmt::format(msg_args...));
-  }
-
+inline void AssertFailed(const char *file_name, int line_num, const char *expr, const std::string &message) {
   spdlog::critical(
       "\nAssertion failed in file {} at line {}."
       "\n\tExpression: '{}'"
       "{}",
-      file_name, line_num, expr, message ? fmt::format("\n\tMessage: '{}'", *message) : "");
+      file_name, line_num, expr, !message.empty() ? fmt::format("\n\tMessage: '{}'", message) : "");
   std::terminate();
 }
 
+#define GET_MESSAGE(...) \
+  BOOST_PP_IF(BOOST_PP_EQUAL(BOOST_PP_VARIADIC_SIZE(__VA_ARGS__), 0), "", fmt::format(__VA_ARGS__))
+
```

Comment:
What about using two functions?

```c++

std::string_view GetMessage() { return ""; }

template <typename... Args>
std::sting GetMessage(const Args&...msg_args) { return fmt::format(msg_args...)); }
```
This way we can avoid using even more macros.

Reply:
Here's the deal:
`fmt::format(msg, args...)` creates `fmt::format_string` from the `msg` argument. That constructor is `consteval` so `msg` should be `constant value`. Arguments of function that is executed during runtime has all the arguments as runtime argument, so the `consteval` requirement will fail. I could avoid that problem by making `AssertFailed` accept `fmt::format_string` as argument (look at `BasicExceptions`), but there is no overload which accepts no msg, so I had to do it with macros.